kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/BigQuerySchemaConverter.java

```diff
@@ -120,33 +119,37 @@ public BigQuerySchemaConverter(boolean allFieldsNullable) {
     return com.google.cloud.bigquery.Schema.of(fields);
   }
 
-  private void throwOnCycle(Schema kafkaConnectSchema, List<Schema> seenSoFar) {
+  private boolean throwOnCycle(Schema kafkaConnectSchema, List<Schema> seenSoFar) {
     if (PRIMITIVE_TYPE_MAP.containsKey(kafkaConnectSchema.type())) {
-      return;
+      return false;
     }
 
     if (seenSoFar.contains(kafkaConnectSchema)) {
-      throw new ConversionConnectException("Kafka Connect schema contains cycle");
+      logger.error("Cycle detected : " + kafkaConnectSchema.name());
```

Comment:
I mention this above, but I don't think this should be a log message (or at least, not exclusively one) and can be included in the message of the exception that ends up being thrown.

Also, what happens if there's a cycle involving nested sub-schemas? Will `kafkaConnectSchema.name()` give us the name of the top-level schema (which is probably what we want to give users), or will it give the name of the nested sub-schema that creates the cycle (which may confuse people as it may not refer to the actual schema they've registered in, e.g., Schema Registry)?

Finally, similar to what you've done with the `LoggerUtils::logRecord` method, should we include metadata about the record that causes this issue in the exception message?

Ultimately, what I think would be useful here is an exception whose message explains that a cycle has been detected in the record schema, contains name of the top-level schema for the record, and has metadata on the topic/partition/offset for the record.

Reply:
Yes. I agree this could be confusing. I've tried to more explicit.

Reply:
We've refactored this to give both the top level schema name as well as the name of the attribute causing the cycle.
Moreover, we also wrapped the `ConnectConversionException` with another one to give details about the specified connect record that causes an issue, so that we didn't need to alter the `SchemaConverter` interface.